framework/src/meshgenerators/MeshDiagnosticsGenerator.C

```diff
@@ -1252,6 +1262,99 @@ MeshDiagnosticsGenerator::checkLocalJacobians(const std::unique_ptr<MeshBase> &
                  num_negative_side_qp_jacobians);
 }
 
+void
+MeshDiagnosticsGenerator::checkNonMatchingEdges(const std::unique_ptr<MeshBase> & mesh) const
+{
+  /*Algorithm Overview
+    1)Prechecks
+      a)This algorithn only works for 3D so check for that first
+    2)Loop
+      a)Loop through every element
+      b)For each element get the edges associated with it
+      c)For each edge check overlap with any edges of nearby elements
+      d)Have check to make sure the same pair of edges are not being tested twice for overlap
+    3)Overlap check
+      a)Shortest line that connects both lines is perpendicular to both lines
+      b)A good overview of the math for finding intersecting lines can be found
+    here->paulbourke.net/geometry/pointlineplane/
+  */
+  if (mesh->mesh_dimension() != 3)
+    mooseError("The edge intersection algorithm only works with 3D meshes");
+  if (!mesh->is_serial())
+    mooseError("Only serialized/replicated meshes are supported");
+  unsigned int num_intersecting_edges = 0;
+  std::set<Node> checked_edges_nodes;
+  for (const auto elem : mesh->active_element_ptr_range())
+  {
+    std::vector<std::unique_ptr<Elem>> elem_edges(elem->n_edges());
+    for (auto i : elem->edge_index_range())
+      elem_edges[i] = elem->build_edge_ptr(i);
+    for (auto other_elem : mesh->active_element_ptr_range())
+    {
+      // If they're the same element then there's no need to check for overlap
+      if (elem == other_elem)
+        continue;
+
+      // Get bounding boxes for both elements. If the bounding boxes don't intersect then no edges
+      // will either
+      auto boundingBox1 = elem->loose_bounding_box();
+      auto boundingBox2 = other_elem->loose_bounding_box();
+      if (!(boundingBox1.intersects(boundingBox2)))
+        continue;
+
+      std::vector<std::unique_ptr<Elem>> other_edges(other_elem->n_edges());
+      for (auto j : other_elem->edge_index_range())
+        other_edges[j] = other_elem->build_edge_ptr(j);
+      for (auto & edge : elem_edges)
+      {
+        for (auto & other_edge : other_edges)
+        {
+          // Get nodes from edges
+          const Node * n1 = edge->get_nodes()[0];
+          const Node * n2 = edge->get_nodes()[1];
+          const Node * n3 = other_edge->get_nodes()[0];
+          const Node * n4 = other_edge->get_nodes()[1];
+
+          // Check if the edges have already been added to our check_edges list
+          if (checked_edges_nodes.count(*n1) && checked_edges_nodes.count(*n2) &&
+              checked_edges_nodes.count(*n3) && checked_edges_nodes.count(*n4))
+          {
+            continue;
+          }
+
+          // Now compare edge with other_edge
+          std::vector<double> intersection_coords(3);
+          bool overlap = MeshBaseDiagnosticsUtils::checkEdgeOverlap(
+              edge, other_edge, intersection_coords, _non_matching_edge_tol);
```

Comment:
this check only works for straight edges btw. EDGE3 will be an issue. add a check on the element type earlier

Reply:
In MeshBaseDiagnosticsUtils I included a mooseAssert at the beginning that checks to see if they are edges. I had it check to make sure they are EDGE2 as well. If you think I should add a check earlier than this I can, but I don't think having the wrong edge type will cause any issues before that check?

Reply:
the assert will not catch user inputs

the assert is great inside the utils, in this generator we ll want a check

Reply:
in the generator, you can tell the user about the type of the parent element, which they ll understand better than edge

Reply:
Sounds good, I'll add a check in the generator then.